Support prefix & suffix for 3 widgets: string, number & boolean

ngdangtu-vn opened the pull request:
This changes are about:
- refactor syntax `BooleanControl` static properties
- add `prefix` & `suffix` options
- update widget control style to match with new options
- update Boolean widget document

<!--
Thanks for submitting a pull request!

Please make sure you've read and understood our contributing guidelines here:
https://github.com/decaporg/decap-cms/blob/master/CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx", where #xxxx is the issue number.

Please provide enough information so that others can review your pull request.
The first two fields are mandatory:
-->

**Summary**

The update allows developers to display messages before (`prefix`) and after (`suffix`) the toggle. This design intends to provide a better information and look to users.

In my use case, I have a field to decide if I want to show a post in homepage. If I set "Show in homepage" string in label, the layout of input would be unbalance as the top small part is bigger than the main body. The toggle itself seems lonely as well. So if I can set "Visibility" in `label` option and "Show in homepage" in `prefix` option, it would be much sense.

**Test plan**

<!--
Demonstrate the code is solid.
Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.
-->

1. Run `yarn start`
2. Go to this path `collections/kitchenSink/entries/a-big-entry-with-all-the-things`
3. Expect to see the exact Boolean input like this screenshot:   
   ![image](https://github.com/decaporg/decap-cms/assets/68631458/ddb01c07-1829-4d48-930e-8d4cb703994e)

**Checklist**

<!-- Please add a `x` inside each checkbox: -->

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).

**A picture of a cute animal (not mandatory but encouraged)**

Be hold! The amazing meowman!

[![meow](https://github.com/decaporg/decap-cms/assets/68631458/5e610ad0-fd00-4a30-9378-3499d103da3d)](https://www.reddit.com/r/Eyebleach/comments/14hfbbe/ʍoǝɯ)

martinjagodic commented:
@ngdangtu-vn I see why you would want this feature, but it seems very similar to `hint`, which exists on all widgets and can be used to explain a field in more detail. Is there a reason why you wouldn't use hint?

ngdangtu-vn commented:
> Is there a reason why you wouldn't use hint?

I wouldn't say I don't use hint. But my reason is simply:

- Not wasting a whole line for only 1 toggle element
- Hint text is small enough to be a hint only. Prefix/Suffix will plays nice the the role of important instruction.
- In some messages, it looks better with prefix (or|and) suffix as well. I admit, my example was very good. Maybe like this: `Show author [x]`.


I even plan to do take this further to `string` and `number` widgets so dev can display fields like this `[...]px`. Cool right?

martinjagodic commented:
Extending this to string and number will be useful! Can you implement that as part of this PR and I will review/merge it all together?

ngdangtu-vn commented:
Update photo for recent commits


![String Widget](https://github.com/decaporg/decap-cms/assets/68631458/ef4e9596-3927-4531-89a1-a60ed16cc49a)

![Number Widget](https://github.com/decaporg/decap-cms/assets/68631458/b4470636-128b-4121-98d8-e93797167c61)

![Boolean Widget](https://github.com/decaporg/decap-cms/assets/68631458/7b1e7e25-0cb2-4de9-9921-14de677093bb)


@martinjagodic Could you review my code and the Widgets' UI? Thanks.

xaiki commented on dev-test/config.yml:
```diff
@@ -117,11 +117,11 @@ collections: # A list of collections the CMS should be able to edit
         value_field: 'title'
-      - { label: 'Title', name: 'title', widget: 'string' }
-      - { label: 'Boolean', name: 'boolean', widget: 'boolean', default: true }
+      - { label: 'Title', name: 'title', widget: 'string', prefix: 'This string:', suffix: 'is a title' }
```
this suffix doesn't seem to make a lot of sense, is it a prefix ? is it a hint ? why would i want this ?

ngdangtu-vn replied:
Nope, it's just for testing and demonstration. But you are welcome to improve it.

xaiki commented on packages/netlify-cms-widget-boolean/src/BooleanControl.js:
```diff
@@ -16,34 +21,42 @@ function BooleanBackground({ isActive, ...props }) {
-
-BooleanControl.defaultProps = {
-  value: false,
-};
```
any reason why these were moved as part of this specific PR ?

ngdangtu-vn replied:
To match with other widget for consistency.

xaiki commented on packages/netlify-cms-widget-number/src/NumberControl.js:
```diff
@@ -100,19 +107,29 @@ export default class NumberControl extends React.Component {
+            onChange={this.handleChange}
+            css={css`flex-grow: 1`}
+          />
+          {suffix && (<span>&nbsp;{suffix}</span>)}
```
wondering if onFocus and onBlur should go on the bigger div (classed {classNameWrapper})

xaiki commented on packages/netlify-cms-widget-string/src/StringControl.js:
```diff
@@ -41,21 +49,30 @@ export default class StringControl extends React.Component {
+            onFocus={setActiveStyle}
+            onBlur={setInactiveStyle}
+            css={css`flex-grow: 1`}
+          />
```
ditto for onFocus/onBlur

xaiki commented on website/content/docs/widgets/boolean.md:
doc is required for number and string widgets too

ngdangtu-vn replied:
Ok, I'm working on it. But I open for your suggestion.

xaiki reviewed (commented):
fwiw, my 2c review

note: i'm not part of the decap team.

ngdangtu-vn commented:
@xaiki

Thanks for your review but I'm not sure if they still need this PR. It kinda a dead PR already.

ngdangtu-vn commented on packages/netlify-cms-widget-boolean/src/BooleanControl.js:
```diff
@@ -16,34 +21,42 @@ function BooleanBackground({ isActive, ...props }) {
+            onBlur={setInactiveStyle}
+            Background={BooleanBackground}
+          />
+          {suffix && (<span>&nbsp;{suffix}</span>)}
```
We can go with `div>span` for both and add `:first-of-type` for prefix, similar goes to suffix. No need to go for class everywhere.

martinjagodic commented:
@xaiki thanks for your reviews

@ngdangtu-vn This PR is not dead, but I understand it sitting here for a long time now. We have quite a backlog and we want to add features based on the triage we decided upon. This one is not essential so it will probably wait a bit longer.

ngdangtu-vn commented:
@martinjagodic Ok it's cool, I thought the code won't be able to fit with the new Decap.

@xaiki So I will fix the PR using your advice later 'cause I'm bit busy recently & I need time to reload my memory :))

xaiki commented:
@martinjagodic if reviews help I'm happy to look at other PRs, just point me to the top of the stack.

martinjagodic commented:
@xaiki I will happily assign you review requests. To talk more in-depth about what to do next, please join the contributing channel on our Discord server.

martinjagodic commented:
@ngdangtu-vn are you still interested in moving this forward? If yes, please address the comments and solve merge conflicts. Thanks!

ngdangtu-vn commented:
Ok, give me around 2 more days

ngdangtu-vn commented:
Hold on, I'm on cleanning up repo.

ngdangtu-vn commented:
I didn't close this? Hold on I'm reorganizing this... Maybe because I reset branch?

martinjagodic commented:
@ngdangtu-vn maybe. I guess you can just reopen and continue from there.

ngdangtu-vn commented on dev-test/config.yml:
```diff
@@ -117,11 +117,11 @@ collections: # A list of collections the CMS should be able to edit
-      - { label: 'Title', name: 'title', widget: 'string' }
-      - { label: 'Boolean', name: 'boolean', widget: 'boolean', default: true }
+      - { label: 'Title', name: 'title', widget: 'string', prefix: 'This string:', suffix: 'is a title' }
+      - { label: 'Boolean', name: 'boolean', widget: 'boolean', prefix: 'Toggle this to switch on/off:', hint: 'Toggle this to switch on/off', default: true }
```
Prefix or Suffix are not only for English, I'm not linguishtic experrt. But I do see many case where other languages would need prefix or suffix. So prefix is needed. And what would hurt the design if I add 1 more feature like prefix, right :)

ngdangtu-vn reviewed (commented):
I'm working merging (copy & paste my old code with modify to fit the new). After that is working on doc example content. I would love to open for any example content suggestion.

ngdangtu-vn commented:
I think I just shot myself to the leg. But it's ok, everything's still under controlled...

ngdangtu-vn committed:
chore(3 widgets): add prefix & suffix labels

Those 3 widgets are Boolean, Number, String

- update 3 widgets UI to display prefix & suffix
- refactor BooleanControl.js code structure
- update document for 3 widgets
- update demo config `dev-test/`

martinjagodic committed:
Merge branch 'master' into widget-boolean

demshy committed:
chore: remove website

demshy committed:
Merge branch 'main' into widget-boolean

demshy committed:
style: lint and update emotion css import

martinjagodic committed:
Merge branch 'main' into widget-boolean

yanthomasdev committed:
Merge branch 'main' into widget-boolean

yanthomasdev committed:
Merge branch 'main' into widget-boolean

yanthomasdev committed:
fix: field type mismatch and missing schema props

yanthomasdev committed:
feat: add tests to widgets

yanthomasdev reviewed (approved):
Thanks @ngdangtu-vn. I made a few changes to your branch, I believe the feature should be in a good place now.

martinjagodic committed:
Merge branch 'main' into widget-boolean

delucis commented:
I’ll note that it might be worth thinking about responsiveness with this change. While `hint` admittedly takes up an additional line, it is responsive in a simple way by breaking up the content vertically and wrapping in an obvious way.

Given a `[prefix] input [suffix]` display, it’s less obvious how that should behave in a narrow viewport: should these wrap onto three lines? only when the content forces it (e.g. `flex-wrap: wrap`? will that layout still look like people expect and convey the prefix/suffix idea to them? (This isn’t only a question for small viewports but also if a user sets a particularly long suffix/prefix.)
